### DISCUSSION TITLE
Python dependency update doesn't respect '<'

### DISCUSSION THREAD

**Discussion by xuhdev** [renovatebot/renovate#10392] [Request Help] [answered]
**How are you running Renovate?**

- [x] WhiteSource Renovate hosted app on github.com
- [ ] Self hosted

**Describe the bug**

For Python pip dependencies, the Renovate bot doesn't respect `<`.

For example, this PR: https://github.com/IBM/MAX-Base/pull/58 . The Renovate bot tried to upgrade `flask>=1.1.2,<2.0` to `flask==2.0.1`. What it should do is to upgrade it to `flask==1.1.4,<2.0`.

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [x] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Comment by rarkins** [comment]
Moved this to a discussion as the issue title was over-generalized and the new constraint is not *wrong*.

**Comment by rarkins** [comment]
Renovate falls back to an exact version constraint if it does not know how to upgrade a complex range. 

>  What it should do is to upgrade it to `flask==1.1.4,<2.0`.

Could you explain why you think this?

I believe that if your existing constraint was `>=1.1.2,<2.0` and the newer version is `2.0.1` then the ideal result is `>=1.1.2,<2.1`.

**Comment by rarkins** [comment, marked as answer]
The `pip_requirements` manager selects `rangeStrategy=pin` if you leave Renovate as the default `rangeStrategy=auto`. This explains why you see the new constraint as `==2.0.1`.

If you want to preserve ranges, you need to add `"rangeStrategy": "replace"` to your Renovate config. Then you'll see `flask>=1.1.2,<2.1` as the new range.
